### PR TITLE
State management refactor

### PR DESCRIPTION
**Pull request by emcelroy**
We've added MobX State Tree to help with state management.

### CODE REVIEW

File: src/hooks/useCodapState.tsx [hunk 1]
```diff
@@ -258,13 +290,40 @@ export const useCodapState = () => {
...
+  const handleSetCollections = (colls: ICollection[]) => {
+    const newCollectionModels = colls.map((coll: ICollection) => {
+      const { areParentChildLinksConfigured, attrs, caseName, cases: _cases, childAttrName,
+              collapseChildren, guid, id, name, parent, title, type } = coll;
+      return {
+        areParentChildLinksConfigured, attrs, caseName, cases: _cases, childAttrName,
+        collapseChildren, guid, id, name, parent, title, type
+      };
+    });
+    applySnapshot(collections, newCollectionModels);
```
Comment (scytacki) [inline src/hooks/useCodapState.tsx:317]: This looks like the same code that is in `updateCollections`, can you consolidate it?

**Review by scytacki — CHANGES_REQUESTED** [review]
This looks really good to me. I left a couple comments about future improvements. 

I think the only thing that would be good to change in this PR is the duplicate code around creating the snapshot to send to `applySnapshot`.

**Comment by emcelroy** [comment]
> This looks really good to me. I left a couple comments about future improvements.
> 
> I think the only thing that would be good to change in this PR is the duplicate code around creating the snapshot to send to `applySnapshot`.

@scytacki  Thanks. I consolidated the duplicate code. I also had a go at updating the collections model like I think you were suggesting. Do those changes make sense? They appear to work fine, though I find the structure a little awkward. It seems a bit weird to have a `collectionsModel` with a `collections` property, but I couldn't think of a better way to do it.

**Comment by kswenson** [comment]
> They appear to work fine, though I find the structure a little awkward. It seems a bit weird to have a `collectionsModel` with a `collections` property, but I couldn't think of a better way to do i

To be more analogous with CODAP itself, `collections` would be a property of the `DataSetModel`.

**Comment by emcelroy** [comment]
> > They appear to work fine, though I find the structure a little awkward. It seems a bit weird to have a `collectionsModel` with a `collections` property, but I couldn't think of a better way to do i
> 
> To be more analogous with CODAP itself, `collections` would be a property of the `DataSetModel`.

@kswenson Thanks. That makes a lot of sense. I spent some time trying to make that work, but it doesn't seem as straightforward as you might expect, and it's still not clear to me how much more time it'd take to work that out. So I'd like to either stick with the current model changes or undo them and save that work for later like Scott originally suggested, if that sounds reasonable to you and @scytacki?

**Review by scytacki — APPROVED** [review]
This looks good to me. Yes I think it is fine to punt the renaming.

There might be some issues with the useEffect dependencies just including `collectionsModel`. However we can probably wait until we find a problem before addressing that.

**Pull request merged**